sys/mathoid.js

```diff
@@ -0,0 +1,141 @@
+'use strict';
+
+
+var P = require('bluebird');
+var HyperSwitch = require('hyperswitch');
+var URI = HyperSwitch.URI;
+var HTTPError = HyperSwitch.HTTPError;
+
+
+function MathoidService(options) {
+
+    this.options = options;
+
+}
+
+
+MathoidService.prototype.checkInput = function(hyper, req) {
+
+    var self = this;
+    var rp = req.params;
+    var hash;
+    var origHash;
+    var checkRes;
+
+    // start by calculating the hash
+    return hyper.post({
+        uri: new URI([rp.domain, 'sys', 'post_data', 'mathoid.input', 'hash']),
+        body: { q: req.body.q, type: rp.type }
+    }).then(function(res) {
+        hash = origHash = res.body;
+        // short-circuit if it's a no-cache request
+        if (req.headers && /no-cache/.test(req.headers['cache-control'])) {
+            return P.reject(new HTTPError({ status: 404 }));
+        }
+        // check the post storage
+        return hyper.get({
+            uri: new URI([rp.domain, 'sys', 'key_value', 'mathoid.check', hash])
+        }).catch({ status: 404 }, function(err) {
+            // let's try to find an indirection
+            return hyper.get({
+                uri: new URI([rp.domain, 'sys', 'key_value', 'mathoid.hash_table', hash])
+            }).then(function(hashRes) {
+                // we have a normalised version of the formula
+                hash = hashRes.body;
+                // grab that version from storage
+                return hyper.get({
+                    uri: new URI([rp.domain, 'sys', 'key_value', 'mathoid.check', hash])
+                });
+            });
+        });
+    }).then(function(res) {
+        // we have a record, return that
+        res.headers['cache-control'] = self.options['cache-control'];
+        return res;
+    }).catch({ status: 404 }, function() {
+        // if we are here, it means this is a new input formula
+        // so call mathoid
+        return hyper.post({
+            uri: self.options.host + '/texvcinfo',
+            headers: { 'content-type': 'application/json' },
+            body: {
+                q: req.body.q,
+                type: rp.type
+            }
+        }).then(function(res) {
+            checkRes = res;
+            // store the normalised version
+            return hyper.put({
+                uri: new URI([rp.domain, 'sys', 'post_data', 'mathoid.input', '']),
```

Comment:
Hm. I'm wondering why are we storing the input at all? Shouldn't we store only the checked/normalized formula?